File: src/CCVTAC.Console/IoUtilties/Directories.cs
using System.IO;
using System.Text;

namespace CCVTAC.Console.IoUtilties;

internal static class Directories
{
    private static readonly string AllFilesSearchPattern = "*";
    private static readonly EnumerationOptions EnumerationOptions = new();

    internal static Result WarnIfHasFiles(string directory, int showMax)
    {
        var fileNames = GetDirectoryFileNames(directory);
        int fileCount = fileNames.Length;

        if (fileNames.IsEmpty)
        {
            return Result.Ok();
        }

        var remainLabel = fileCount == 1 ? "file remains" : "files remain";
        var report = new StringBuilder($"{fileCount} {remainLabel} in working directory \"{directory}\":");

        foreach (string fileName in fileNames.Take(showMax))
        {
            report.AppendLine($"• {fileName}");
        }

        if (fileCount > showMax)
        {
            report.AppendLine($"... plus {fileCount - showMax} more.");
        }

        return Result.Fail(report.ToString());
    }

    /// <summary>
    /// Returns the filenames in a given directory, optionally ignoring specific filenames.
    /// </summary>
    /// <param name="directoryName"></param>
    /// <param name="customIgnoreFiles">An optional list of files to be excluded.</param>
    private static ImmutableArray<string> GetDirectoryFileNames(
        string directoryName,
        IEnumerable<string>? customIgnoreFiles = null)
    {
        var ignoreFiles = customIgnoreFiles?.Distinct() ?? [];

        return
            [
                ..Directory
                    .GetFiles(directoryName, AllFilesSearchPattern, EnumerationOptions)
                    .Where(filePath =>
                        ignoreFiles.None(ignoreFile => filePath.EndsWith(ignoreFile)))
            ];
    }
}

File: src/CCVTAC.Main/IoUtilities/Directories.fs
namespace CCVTAC.Main.IoUtilities

open CCVTAC.Main
open System.IO

module Directories =

    [<Literal>]
    let private allFilesSearchPattern = "*"

    /// Counts the number of audio files in a directory.
    let audioFileCount (directory: string) (includedExtensions: string list) =
        DirectoryInfo(directory).EnumerateFiles()
        |> Seq.filter (fun f -> List.caseInsensitiveContains f.Extension includedExtensions)
        |> Seq.length

    /// Returns the filenames in a given directory, optionally ignoring specific filenames.
    let private getDirectoryFileNames
        (directoryName: string)
        (customIgnoreFiles: string seq option)
        : Result<string array, string> =

        let ignoreFiles =
            customIgnoreFiles
            |> Option.defaultValue Seq.empty
            |> Seq.distinct
            |> Seq.toArray

        ofTry (fun _ ->
            Directory.GetFiles(directoryName, allFilesSearchPattern, EnumerationOptions())
            |> Array.filter (fun filePath -> not (ignoreFiles |> Array.exists filePath.EndsWith)))

    let deleteAllFiles workingDirectory
        : Result<ResultMessageCollection, string> =

        let delete fileNames =
            let successes, failures = ResizeArray<string>(), ResizeArray<string>()

            for fileName in fileNames do
                try
                    File.Delete fileName
                    successes.Add $"• Deleted \"%s{fileName}\""
                with exn ->
                    failures.Add $"• Error deleting \"%s{fileName}\": %s{exn.Message}"

            { Successes = successes |> Seq.toList |> List.rev
              Failures  = failures  |> Seq.toList |> List.rev }

        match getDirectoryFileNames workingDirectory None with
        | Error errMsg -> Error errMsg
        | Ok fileNames -> Ok (delete fileNames)

    /// Ask the user to confirm the deletion of files in the specified directory.
    let askToDeleteAllFiles dirName (printer: Printer) =
        if printer.AskToBool("Delete all temporary files?", "Yes", "No")
        then deleteAllFiles dirName
        else Error "Will not delete the files."

    let printDeletionResults (printer: Printer) (results: ResultMessageCollection) : unit =
        printer.Info $"Deleted %s{String.fileLabel results.Successes.Length}."
        results.Successes |> List.iter printer.Debug

        if List.isNotEmpty results.Failures then
            printer.Warning $"However, %s{String.fileLabel results.Failures.Length} could not be deleted:"
            results.Failures |> List.iter printer.Error

    let warnIfAnyFiles showMax dirName =
        match getDirectoryFileNames dirName None with
        | Error errMsg -> Error errMsg
        | Ok fileNames ->
            if Array.isEmpty fileNames then
                Ok ()
            else
                SB($"Unexpectedly found {String.fileLabel fileNames.Length} in working directory \"{dirName}\":{String.newLine}")
                    .AppendLine
                        (fileNames
                         |> Array.truncate showMax
                         |> Array.map (sprintf "• %s")
                         |> String.concat String.newLine)
                |> fun sb ->
                    if fileNames.Length > showMax
                    then sb.AppendLine $"... plus {fileNames.Length - showMax} more."
                    else sb
                |> _.AppendLine("This sometimes occurs due to the same video appearing twice in playlists.")
                |> _.ToString()
                |> Error

    /// Ensures the specified directory exists, including creation of it if necessary.
    let ensureDirectoryExists dirName : Result<DirectoryInfo, string> =
        try
            dirName
            |> Path.GetFullPath
            |> Directory.CreateDirectory
            |> Ok
        with exn ->
            Error $"Error accessing or creating directory \"%s{dirName}\": %s{exn.Message}"

